File: api/data/NMBS/tools/VehicleIdTools.php
<?php


class VehicleIdTools
{
    /**
     * Extract the train number from a vehicle id. For example, S104522 will return 4522.
     * @param string $vehicleId
     * @return string
     */
    public static function extractTrainNumber(string $vehicleId): string
    {
        $vehicleId = strtoupper($vehicleId);
        // Handle S trains. For example, S5 3381 or S53381 should become 3381. Typically a number has 4 digits.
        $vehicleId = preg_replace("/S[12]0 ?(\d{4})/", "$1", $vehicleId); // S10, S20
        $vehicleId = preg_replace("/S3[234] ?(\d{4})/", "$1", $vehicleId); // S32, S33, S34
        $vehicleId = preg_replace("/S4[1234] ?(\d{4})/", "$1", $vehicleId); // S41, 42, 43, 44
        $vehicleId = preg_replace("/S51 ?(7\d{2})/", "$1", $vehicleId); // S51 750, 751, ...
        $vehicleId = preg_replace("/S52 ?(\d{4})/", "$1", $vehicleId); // S51, 52, 53 (those often have 3-digit numbers)
        $vehicleId = preg_replace("/S53 ?(6\d{2})/", "$1", $vehicleId); // S53 650, 651, ...
        $vehicleId = preg_replace("/S6[1234] ?(\d{4})/", "$1", $vehicleId); // S61, 62, 63, 64
        $vehicleId = preg_replace("/S81 ?([78]\d{4})/", "$1", $vehicleId); // S81 7xxx or 8xxx
        $vehicleId = preg_replace("/S[0-9] ?/", "", $vehicleId); // S1-S9
        $vehicleId = preg_replace("/[^0-9]/", "", $vehicleId);
        return $vehicleId;
    }

    public static function extractTrainType(string $vehicleId): string
    {
        return trim(substr($vehicleId, 0, -strlen(self::extractTrainNumber($vehicleId))));
    }
}

File: src/main/java/be/irail/api/util/VehicleIdTools.java
package be.irail.api.util;

import be.irail.api.exception.InternalProcessingException;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

/**
 * Utility class for parsing and extracting information from vehicle identifiers.
 * Handles Belgian train numbering schemes, including S-trains and buses.
 */
public class VehicleIdTools {
    private static final Logger log = LogManager.getLogger(VehicleIdTools.class);

    /**
     * Extracts the numeric train number from a full vehicle ID.
     * For example, "S104522" will return "4522".
     *
     * @param vehicleId the full vehicle ID (e.g., "IC 548", "S104522")
     * @return the extracted train number as a string
     */
    public static int extractTrainNumber(String vehicleId) {
        String id = extractTrainNumberString(vehicleId);
        int result = 0;
        try {
            result = Integer.parseInt(id);
        } catch (NumberFormatException e) {
            log.error("Failed to extract journey number for vehicle " + vehicleId);
            throw new InternalProcessingException("Failed to extract journey number for vehicle " + vehicleId);
        }
        return result;
    }

    private static String extractTrainNumberString(String vehicleId) {
        if (vehicleId == null) {
            return null;
        }
        String id = vehicleId.toUpperCase();
        // Handle S trains. For example, S5 3381 or S53381 should become 3381. Typically a number has 4 digits.
        id = id.replaceAll("BUS ?(\\d{3,5})", "$1"); // BUSS123, BUSS23456
        id = id.replaceAll("S[12]0 ?(\\d{4})", "$1"); // S10, S20
        id = id.replaceAll("S19 ?(\\d{4})", "$1"); // S19
        id = id.replaceAll("S32 ?(4\\d{2})", "$1"); // S32 400
        id = id.replaceAll("S3[2345] ?(\\d{4})", "$1"); // S32, S33, S34
        id = id.replaceAll("S4[1234] ?(\\d{4})", "$1"); // S41, 42, 43, 44
        id = id.replaceAll("S51 ?(\\d{3,4})", "$1"); // S51
        id = id.replaceAll("S52 ?(\\d{4})", "$1"); // S51, 52, 53
        id = id.replaceAll("S53 ?(6\\d{2})", "$1"); // S53 650, 651, ...
        id = id.replaceAll("S6[1234] ?(\\d{4})", "$1"); // S61, 62, 63, 64
        id = id.replaceAll("S81 ?([78]\\d{4})", "$1"); // S81 7xxx or 8xxx
        id = id.replaceAll("S[0-9] ?", ""); // S1-S9
        id = id.replaceAll("[^0-9]", "");
        return id;
    }

    /**
     * Extracts the train type from a full vehicle ID.
     * For example, "S104522" will return "S10".
     *
     * @param vehicleId the full vehicle ID
     * @return the extracted train type (e.g., "IC", "S10")
     */
    public static String extractTrainType(String vehicleId) {
        if (vehicleId == null) {
            return null;
        }
        String number = extractTrainNumberString(vehicleId);
        return vehicleId.substring(0, vehicleId.length() - number.length()).trim();
    }
}
